### ISSUE TITLE
Don't pass lambda function pointers to external code

### ISSUE THREAD

**Issue by sdilts** [carp-lang/Carp#840]
When passing a  lambda to an external function, the compiler will pass the function pointer held within the `Lambda` struct as an argument, and not the entire lambda struct. While this might work when a lambda function doesn't capture any variables, it will always fail when variables are captured, as the env struct will be dropped an the number of arguments to the lambda function will be incorrect.

Here is an example:
``` clojure
(register callback (Fn [(Fn [Int] Int)] Int))

(defn call-callback [val]
  (let [func (Fn [x] (+ x val))]
     (callback func)))
```
The following C code will be generated:
``` C
int _Lambda_call_MINUS_callback_13(_Lambda_call_MINUS_callback_13_env* _env, int x) {
    int _12 = Int__PLUS_(x, _env->val);
    return _12;
}

int call_MINUS_callback(int val) {
    int _18;
    /* let */ {
        // This lambda captures 1 variables: val
        _Lambda_call_MINUS_callback_13_env *_13_env = CARP_MALLOC(sizeof(_Lambda_call_MINUS_callback_13_env));
        _13_env->val = val;
        Lambda _13 = {
          .callback = _Lambda_call_MINUS_callback_13,
          .env = _13_env,
          .delete = _Lambda_call_MINUS_callback_13_env_delete,
          .copy = _Lambda_call_MINUS_callback_13_env_copy
        };
        Lambda func = _13;
        int _17 = callback(func.callback);
        _18 = _17;
    }
    return _18;
}
```

While the `callback` function expects to be given a function with one argument, it is given an function that takes two arguments, which results in undefined behavior.

**Comment by sdilts** [comment]
A potential fix would be to pass the `Lambda` struct to the external function, and expect it to know what to do with it.

**Comment by sdilts** [comment]
Passing a reference to a lambda will result in a pointer to the `Lambda` struct being passed as an argument, so that is what should be used instead.

**Comment by eriksvedang** [comment]
> A potential fix would be to pass the `Lambda` struct to the external function, and expect it to know what to do with it.

Sounds like a good idea. We could even have some C functions/macros available to help the C programmer...

**Comment by jacereda** [comment]
C APIs normally expose callback registration by provinding a second `void *` argument for receiving some information in the callback. Doesn't the code you're trying to wrap provide such `void *` argument?

**Comment by eriksvedang** [comment]
At the very least, we should probably document this behaviour somewhere.

**Comment by TimDeve** [comment]
So I've been thinking about how we can make it easier to interface with C functions that have the signature that @jacereda describes, I thought we could add some functions that returns the pointers to the function and env in lambdas. [This is what I came up with.](https://github.com/TimDeve/unsafe-fn-ptr/blob/master/unsafe-fn-ptr.carp) I couldn't really think of a better signature that (Ref a) for the function but my thinking would be that these would be wrapped in something safer like the [example for threads](https://github.com/TimDeve/unsafe-fn-ptr/blob/master/examples/thread/main.carp) I have in the repo.

**Comment by eriksvedang** [comment]
@TimDeve Looks great!

One question: When do you anticipate needing to use the `to-void-ptr` function? Is that for sending the function pointer to some code that actually knows about how we organize lambdas in Carp?

We should probably put them in the existing `Pointer` module (we could actually generate one of each of these for different arities, like with `copy` etc to make sure that you're passing a function and not anything).

Finally we would need some tests, since these things are kinda tricky to get right :)

**Comment by TimDeve** [comment]
`to-void-pointer` would be if you want to pass an argument to a function like [here](https://github.com/TimDeve/unsafe-fn-ptr/blob/345b372815b506acbfd10d2616bea0b30dc2ba81/examples/basic/main.carp#L22).

**Comment by eriksvedang** [comment]
Perhaps this can be closed by #1038 ?

**Comment by TimDeve** [comment]
I think it should be enough to close this issue.
